primitives/src/xcm.rs

```diff
@@ -312,3 +320,117 @@ where
+        AccountIdConverter,
+        NonNativeMatcher,
+        MultiAdapterFungibleLedger,
+        MultiAdapterAssetConfig,
+    )>,
+);
+
+impl<
+        Runtime: frame_system::Config,
+        AccountIdConverter: XcmConvert<MultiLocation, Runtime::AccountId>,
+        NativeMatcher: MatchesFungible<Balance>,
+        NonNativeMatcher: MatchesFungibles<AssetId, Balance>,
+        MultiAdapterFungibleLedger: FungibleLedger<Runtime>,
+        MultiAdapterAssetConfig: AssetConfig<Runtime>,
+    > TransactAsset
+    for MultiAssetAdapter<
+        Runtime,
+        AccountIdConverter,
+        NativeMatcher,
+        NonNativeMatcher,
+        MultiAdapterFungibleLedger,
+        MultiAdapterAssetConfig,
+    >
+{
+    fn deposit_asset(asset: &MultiAsset, location: &MultiLocation) -> XcmResult {
+        log::trace!(
+            target: "xcm::multi_asset_adapter",
+            "deposit_asset asset: {:?}, location: {:?}",
+            asset, location,
+        );
+
+        let (asset_id, amount, who) = Self::match_asset_and_location(asset, location)?;
+
+        MultiAdapterFungibleLedger::mint(asset_id, &who, amount)
+            .map_err(|_| XcmError::FailedToTransactAsset("Failed Mint"))?;
+
+        Ok(())
+    }
+
+    fn withdraw_asset(
+        asset: &MultiAsset,
+        location: &MultiLocation,
+    ) -> result::Result<xcm_executor::Assets, XcmError> {
+        log::trace!(
+            target: "xcm::multi_asset_adapter",
+            "withdraw_asset asset: {:?}, location: {:?}",
+            asset, location,
+        );
+
+        let (asset_id, amount, who) = Self::match_asset_and_location(asset, location)?;
+
+        MultiAdapterFungibleLedger::burn(asset_id, &who, amount)
+            .map_err(|_| XcmError::FailedToTransactAsset("Failed Burn"))?;
+
+        Ok(asset.clone().into())
+    }
+}
+
+impl<
+        Runtime: frame_system::Config,
+        AccountIdConverter: XcmConvert<MultiLocation, Runtime::AccountId>,
+        NativeMatcher: MatchesFungible<Balance>,
+        NonNativeMatcher: MatchesFungibles<AssetId, Balance>,
+        MultiAdapterFungibleLedger: FungibleLedger<Runtime>,
+        MultiAdapterAssetConfig: AssetConfig<Runtime>,
+    >
+    MultiAssetAdapter<
+        Runtime,
+        AccountIdConverter,
+        NativeMatcher,
+        NonNativeMatcher,
+        MultiAdapterFungibleLedger,
+        MultiAdapterAssetConfig,
+    >
+{
+    /// Matches the incoming `asset` to an `asset_id` and `amount` on this chain.
+    /// Matches the incoming `location` to a `receiver` account on this chain.
+    /// Uses the matcher implementation of both native and non-native assets.
+    /// Returns the `asset_id`, `amount` and `receiver` if all three were matched.
+    fn match_asset_and_location(
+        asset: &MultiAsset,
+        location: &MultiLocation,
+    ) -> result::Result<(AssetId, Balance, Runtime::AccountId), XcmError> {
+        let receiver = AccountIdConverter::convert_ref(location).map_err(|_| {
```

Comment:
Does it work for account type like `AccountKey20`?

Reply:
Depends on the runtime configuration. In our particular case we don't handle `AccountKey20` locations. But remember that the expected sender origin is actually the parachain sovereign account, not a user account. That's why we don't need to handle this conversion.
User account origin logic is handled on the sender side. Once the message is passed to the backend transportation system, the origin is changed to the parachain id of the sender.